// redis-modules-sdk/lib/modules/module.base.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Module = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */
const IORedis = require("ioredis");
class Module {
    constructor(name, options, moduleOptions, clusterOptions) {
        this.name = name;
        //If it's a list of cluster nodes
        if (Array.isArray(options))
            this.clusterNodes = options;
        else
            this.redisOptions = options;
        this.isHandleError = moduleOptions && moduleOptions.isHandleError ? moduleOptions.isHandleError : true;
        this.showDebugLogs = moduleOptions && moduleOptions.showDebugLogs ? moduleOptions.showDebugLogs : false;
        this.clusterOptions = clusterOptions ? clusterOptions : undefined;
    }
    /**
     * Connecting to the Redis database with the module
     */
    connect() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.clusterNodes)
                this.cluster = new IORedis.Cluster(this.clusterNodes, this.clusterOptions);
            else
                this.redis = new IORedis(this.redisOptions);
        });
    }
    /**
     * Disconnecting from the Redis database with the module
     */
    disconnect() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.clusterNodes)
                yield this.cluster.quit();
            else
                yield this.redis.quit();
        });
    }
    /**
     * Running a Redis command
     * @param command The redis command
     * @param args The args of the redis command
     */
    sendCommand(command, args = []) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (this.showDebugLogs)
                    console.log(`${this.name}: Running command ${command} with arguments: ${args}`);
                const response = this.clusterNodes ? yield this.cluster.cluster.call(command, args) : yield this.redis.send_command(command, args);
                if (this.showDebugLogs)
                    console.log(`${this.name}: command ${command} responded with ${response}`);
                return response;
            }
            catch (error) {
                return this.handleError(`${this.name} class (${command.split(' ')[0]}): ${error}`);
            }
        });
    }
    /**
     * Handling a error
     * @param module The name of the module
     * @param error The message of the error
     */
    handleError(error) {
        if (this.isHandleError)
            throw new Error(error);
        return error;
    }
    /**
     * Simpilizing the response of the Module command
     * @param response The array response from the module
     */
    handleResponse(response) {
        const obj = {};
        //If not an array/object
        if (typeof response === 'string' ||
            typeof response === 'number' ||
            (Array.isArray(response) && response.length % 2 === 1 && response.length > 1 && !this.isOnlyTwoDimensionalArray(response)) ||
            (Array.isArray(response) && response.length === 0))
            return response;
        else if (Array.isArray(response) && response.length === 1)
            return this.handleResponse(response[0]);
        else if (Array.isArray(response) && response.length > 1 && this.isOnlyTwoDimensionalArray(response))
            return this.handleResponse(this.reduceArrayDimension(response));
        //If is an array/obj we will build it
        for (let i = 0; i < response.length; i += 2) {
            if (response[i + 1] !== '' && response[i + 1] !== undefined) {
                if (Array.isArray(response[i + 1]) && this.isOnlyTwoDimensionalArray(response[i + 1])) {
                    obj[response[i]] = this.reduceArrayDimension(response[i + 1]);
                    continue;
                }
                const value = (Array.isArray(response[i + 1]) ? this.handleResponse(response[i + 1]) : response[i + 1]);
                obj[response[i]] = value;
            }
        }
        return obj;
    }
    /**
     * Check if array is fully two dimensional. Only items in the array are arrays.
     * @param array The potential two dimensional array
     */
    isOnlyTwoDimensionalArray(array) {
        return array.filter(item => Array.isArray(item)).length === array.length;
    }
    /**
     * Reducing an array by one level. i.e. from two dimensional to 1 dimensional.
     * @param array The potentional two dimensional array
     */
    reduceArrayDimension(array) {
        let newArray = [];
        array.forEach(singleArr => {
            newArray = newArray.concat(singleArr);
        });
        return newArray;
    }
}
exports.Module = Module;
