// redis-modules-sdk/lib/modules/redisgears.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisGears = void 0;
const module_base_1 = require("./module.base");
class RedisGears extends module_base_1.Module {
    constructor(options, moduleOptions, clusterOptions) {
        super(RedisGears.name, options, moduleOptions, clusterOptions);
    }
    /**
     * Aborting an existing execution
     * @param id The id of the execution
     */
    abortExecution(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('RG.ABORTEXECUTION', [id]);
        });
    }
    /**
     * Retrieving key's configuration
     * @param key A list of keys
     */
    configGet(key) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('RG.CONFIGGET', key);
        });
    }
    /**
     * Setting key's configuration
     * @param keyvalue A key value array, i.e. [['key', 'value]]
     */
    configSet(keyvalues) {
        return __awaiter(this, void 0, void 0, function* () {
            const args = [];
            for (const keyvalue of keyvalues)
                args.concat(keyvalue);
            return yield this.sendCommand('RG.CONFIGSET', args);
        });
    }
    /**
     * Dropping an existing execution
     * @param id The id of the execution
     */
    dropExecution(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('RG.DROPEXECUTION', [id]);
        });
    }
    /**
     * Dumping all of the executions
     */
    dumpExecutions() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('RG.DUMPEXECUTIONS');
        });
    }
    /**
     * Dumping all of the registrations
     */
    dumpRegistrations() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('RG.DUMPREGISTRATIONS');
        });
    }
    /**
     * Retrieving an execution
     * @param id The id of the execution
     * @param options The additional optional parameters
     */
    getExecution(id, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const args = [id.toString()];
            if (options !== undefined && options.shard === true)
                args.push('SHARD');
            if (options !== undefined && options.cluster === true)
                args.push('CLUSTER');
            return yield this.sendCommand('RG.GETEXECUTION', args);
        });
    }
    /**
     * Retrieving the results
     * @param id The id of the execution
     */
    getResults(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('RG.GETRESULTS', [id]);
        });
    }
    /**
     * Retrieving the results that have 'UNBLOCKING' argument (And removing it)
     * @param id The id of the execution
     */
    getResultsBlocking(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('RG.GETRESULTSBLOCKING', [id]);
        });
    }
    /**
     * Retrieving information about the cluster
     */
    infocluster() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('RG.INFOCLUSTER');
        });
    }
    /**
     * Executing a python function
     * @param func The function
     * @param options The additional optional arguments
     */
    pyexecute(func, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const args = [func];
            if (options !== undefined && options.unblocking === true)
                args.push('UNBLOCKING');
            if (options !== undefined && options.requirements !== undefined)
                args.concat(['REQUIREMENTS'].concat(options.requirements));
            return yield this.sendCommand('RG.PYEXECUTE', args);
        });
    }
    /**
     * Retrieving memory usage statistics from the 'Python interpreter'
     */
    pystats() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('RG.PYSTATS');
        });
    }
    /**
     * Retrieving a list of all the python requirements available
     */
    pydumpreqs() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('RG.PYDUMPREQS');
        });
    }
    /**
     * Refreshing the node's view of the cluster's topology
     */
    refreshCluster() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('RG.REFRESHCLUSTER');
        });
    }
    /**
     * Triggering the execution of a registered 'CommandReader' function
     * @param trigger The trigger's name
     * @param args The additional arguments
     */
    trigger(trigger, args) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('RG.TRIGGER', [trigger].concat(args));
        });
    }
    /**
     * Removing the registration of a function
     * @param id The id of the execution
     */
    unregister(id) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('RG.UNREGISTER', [id]);
        });
    }
}
exports.RedisGears = RedisGears;
