// redis-modules-sdk/lib/modules/ris.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisIntervalSets = void 0;
const module_base_1 = require("./module.base");
class RedisIntervalSets extends module_base_1.Module {
    constructor(options, moduleOptions, clusterOptions) {
        super(RedisIntervalSets.name, options, moduleOptions, clusterOptions);
    }
    /**
     * Adding an interval set
     * @param key The name of the key
     * @param sets A list of sets to create. At least 1 set is required.
     */
    add(key, sets) {
        return __awaiter(this, void 0, void 0, function* () {
            let args = [key];
            for (const set of sets)
                args = args.concat([set.name, set.minimum, set.maximum]);
            return yield this.sendCommand('iset.add', args);
        });
    }
    /**
     * Retrieving all of key interval sets/a single set.
     * @param key The name of the key
     * @param setName Optional. The name of specific set. If not passed all interval sets under key will be retrieved.
     */
    get(key, setName) {
        return __awaiter(this, void 0, void 0, function* () {
            const args = [key];
            if (setName)
                args.push(setName);
            const response = yield this.sendCommand('iset.get', args);
            return this.parseGet(response);
        });
    }
    /**
     * Deleting a all interval sets under a key, or a single/list of specific set/s.
     * @param key The name of the key
     * @param setNames Optional. A list of set names to delete. If not passed all interval sets under key will be removed.
     */
    del(key, setNames) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('iset.del', [key].concat(setNames));
        });
    }
    /**
     * Retrieving all sets under a key that have a specific score in their range.
     * @param key The name of the key
     * @param score The score of the set
     */
    score(key, score) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('iset.score', [key, score]);
        });
    }
    /**
     * Retrieving all sets under a key that don't have a specific score in their range.
     * @param key The name of the key
     * @param score The score of the set
     */
    notScore(key, score) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.sendCommand('iset.not_score', [key, score]);
        });
    }
    /**
     * Parsing the iset.get command response
     * @param sets The list of sets
     */
    parseGet(sets) {
        const parsedSets = [];
        for (const set of sets) {
            if (set.length > 2)
                parsedSets.push({ name: set[0], minimum: parseInt(set[1]), maximum: parseInt(set[2]) });
            else
                return [{ minimum: parseInt(set[0]), maximum: parseInt(set[1]) }];
        }
        return parsedSets;
    }
}
exports.RedisIntervalSets = RedisIntervalSets;
